ktor-io/common/src/io/ktor/utils/io/ByteReadChannelOperations.kt

```diff
@@ -372,62 +374,41 @@ public suspend fun ByteReadChannel.readUTF8LineTo(out: Appendable, max: Int = In
     if (readBuffer.exhausted()) awaitContent()
     if (isClosedForRead) return false
 
-    var consumed = 0
-    while (!isClosedForRead) {
-        awaitContent()
-
-        val cr = readBuffer.indexOf('\r'.code.toByte())
-        val lf = readBuffer.indexOf('\n'.code.toByte())
-
-        // No new line separator
-        if (cr == -1L && lf == -1L) {
-            if (max == Int.MAX_VALUE) {
-                val value = readBuffer.readString()
-                out.append(value)
-            } else {
-                val count = minOf(max - consumed, readBuffer.remaining.toInt())
-                consumed += count
-                out.append(readBuffer.readString(count.toLong()))
-
-                if (consumed == max) throw TooLongLineException("Line exceeds limit of $max characters")
+    Buffer().use { lineBuffer ->
+        while (!isClosedForRead) {
+            while (!readBuffer.exhausted()) {
+                when (val b = readBuffer.readByte()) {
+                    CR -> {
```

Comment:
Yeah, it's all buffered anyway so it seems to be roughly the same implementation.

Oh!  I just realized why it's so slow!  It's because if you have an input with no `\r` characters, then it would be O(n^2) for the previous implementation, because it searches to the end of the content every time!

Reply:
The `readBuffer.indexOf('\r'.code.toByte())` call is problematic if the input doesn't contain any `\r` characters. Then it always scans till the end for each `readUTF8LineTo` call.
The time spent for reading all lines doesn't grow linearly, but with `O(n²)`. For even larger inputs it's more than 100 times slower.